### PR TITLE
[DCOS-41589][Integration Tests] --supervise causing frameworkId conflicts in mesos cluster mode

### PR DESCRIPTION
**Pull request by farhan5900**
## What changes were proposed in this pull request?

Resolves [DCOS-41589](https://jira.mesosphere.com/browse/DCOS-41589)

Added an integration test to check the supervise mode of the spark application as well as it's conflicts of framework Id's, when job is restarting.

## How were these changes tested?

It was tested running manually on the local machine.

### CODE REVIEW

File: tests/test_spark.py [hunk 1]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
         assert stdout_file["size"] > 0, "stdout file should have content"
     finally:
         sdk_install.uninstall(utils.SPARK_PACKAGE_NAME, service_name)
+
+
+@pytest.mark.sanity
+def test_supervise_conflict_frameworkid():
```
Comment (samvantran) [inline tests/test_spark.py]: overall, the test code looks good but teamcity CI fails because it looks you're trying to submit a spark job after the dispatcher has already been uninstalled [in the previous test](https://github.com/mesosphere/spark-build/pull/506/files#diff-2caedadc3f4e69445462b60b70423fcfR391). Suggest either reusing a previous dispatcher (saves time) or redeploying one for your test

File: tests/test_spark.py [hunk 2]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+        svc = shakedown.get_service(job_service_name)
+        if present:
+            return svc is not None
+        else:
+            return svc is None
+
+    job_args = ["--supervise",
+                "--class", "MockTaskRunner",
+                "--conf", "spark.cores.max=8",
+                "--conf", "spark.executors.cores=4"]
```
Comment (akirillov) [inline tests/test_spark.py]: let's be more conservative and use 1 executor with 1 core. When we move to parallel test execution it will save other tests from resource starvation.

File: tests/test_spark.py [hunk 3]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+
+    driver_id = utils.submit_job(app_url=utils.dcos_test_jar_url(),
```
Comment (akirillov) [inline tests/test_spark.py]: after submitting a driver, we need to wrap the following code in try-finally and kill the Driver when the test is over or failed in the middle. It'll save us from orphaned jobs in the cluster.

This pattern occurs more often so we'll address it separately and extract this handling into utils as a separate PR.

File: tests/test_spark.py [hunk 4]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+                                 app_args="4 1800",
+                                 service_name=utils.SPARK_SERVICE_NAME,
+                                 args=job_args)
+    LOGGER.info("Started supervised driver {}".format(driver_id))
+    wait_job_present(True)
+    LOGGER.info("Job has registered")
+    sdk_tasks.check_running(job_service_name, 1)
+    LOGGER.info("Job has running executors")
```
Comment (akirillov) [inline tests/test_spark.py]: can we please rename `LOGGER` into `log` like in all newly developed tests

File: tests/test_spark.py [hunk 5]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+
+    service_info = shakedown.get_service(job_service_name).dict()
+    driver_regex = "spark.mesos.driver.frameworkId={}".format(service_info['id'])
+
+    status, stdout = shakedown.run_command_on_agent(service_info['hostname'],
```
Comment (akirillov) [inline tests/test_spark.py]: while this is a valid approach, it has a drawback of multiple ssh trips to the agent host. I'd suggest using task exec instead, e.g.
```
dcos task exec <submission ID> "ps aux | grep java | <clear the noise with something like 'cut d' ' -f1'> | xargs kill -9
```

File: tests/test_spark.py [hunk 6]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+        status, stdout = shakedown.run_command_on_agent(service_info['hostname'],
+                                                        "sudo kill -9 {}".format(pid),
+                                                        username=sdk_cmd.LINUX_USER)
+
+        if status:
+            print("Killed pid: {}".format(pid))
+        else:
+            print("Unable to killed pid: {}".format(pid))
+
+    wait_job_present(True)
```
Comment (akirillov) [inline tests/test_spark.py]: before that, we need to verify that the original task has been stopped/killed

File: tests/test_spark.py [hunk 7]
```diff
@@ -389,3 +389,67 @@ def test_task_stdout():
...
+    LOGGER.info("Job has re-registered")
+    sdk_tasks.check_running(job_service_name, 1)
+    LOGGER.info("Job has re-started")
+
+    restarted_service_info = shakedown.get_service(job_service_name).dict()
+    assert service_info['id'] != restarted_service_info['id'], "Job has restarted with same framework Id"
+
+    kill_info = utils.kill_driver(driver_id, utils.SPARK_SERVICE_NAME)
```
Comment (akirillov) [inline tests/test_spark.py]: great that we check for valid termination as well 👍

File: tests/test_spark.py (hunk 1 above)
Comment (farhan5900) [reply tests/test_spark.py]: As far as I understood, in this module dispatcher is auto installed, it will be reused. The teamcity CI fails because of [this](https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1800891&_focus=19929) reason.

File: tests/test_spark.py (hunk 2 above)
Comment (farhan5900) [reply tests/test_spark.py]: Reduced it to 1.

File: tests/test_spark.py (hunk 3 above)
Comment (farhan5900) [reply tests/test_spark.py]: Added try-finally block.

File: tests/test_spark.py (hunk 4 above)
Comment (farhan5900) [reply tests/test_spark.py]: Renamed `LOGGER` to `log`.

File: tests/test_spark.py (hunk 6 above)
Comment (farhan5900) [reply tests/test_spark.py]: Added statement to check whether job is stopped/killed.

File: tests/test_spark.py (hunk 7 above)
Comment (farhan5900) [reply tests/test_spark.py]: Thanks :+1:

File: tests/test_spark.py (hunk 5 above)
Comment (farhan5900) [reply tests/test_spark.py]: `dcos task exec` is not working in this case it is saying `Command is allowed only for UCR`.
Instead of doing multiple ssh, changed it to make single ssh and kill the process.

File: tests/test_spark.py (hunk 1 above)
Comment (samvantran) [reply tests/test_spark.py]: I saw [this error](https://teamcity.mesosphere.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1800891&_focus=18660) which caused the diag bundle collection due to failure.

File: tests/test_spark.py [hunk 8]
```diff
@@ -389,3 +389,64 @@ def test_task_stdout():
...
+
+        wait_job_present(True)
+        log.info("Job has registered")
+
+        sdk_tasks.check_running(job_service_name, 1)
+        log.info("Job has running executors")
+
+        service_info = shakedown.get_service(job_service_name).dict()
+        driver_regex = "spark.mesos.driver.frameworkId={}".format(service_info['id'])
+        status, stdout = shakedown.run_command_on_agent(service_info['hostname'],
```
Comment (akirillov) [inline tests/test_spark.py]: now it looks way cleaner, thanks. Keep in mind that this call causes SSH to an agent and given the known issue with `shakedown` might cause the sporadic failure of this or some irrelevant tests. I'd highly recommend switching to `dcos task exec` instead.

Also, we have a similar method implemented in `sdk_cmd.kill_task_with_pattern` - let's use that one instead

File: tests/test_spark.py [hunk 9]
```diff
@@ -389,3 +389,64 @@ def test_task_stdout():
...
+        log.info("Started supervised driver {}".format(driver_id))
+
+        wait_job_present(True)
+        log.info("Job has registered")
+
+        sdk_tasks.check_running(job_service_name, 1)
+        log.info("Job has running executors")
+
+        service_info = shakedown.get_service(job_service_name).dict()
+        driver_regex = "spark.mesos.driver.frameworkId={}".format(service_info['id'])
```
Comment (akirillov) [inline tests/test_spark.py]: is `spark.mesos.driver.frameworkId=` part of regexp necessary?

File: tests/test_spark.py [hunk 10]
```diff
@@ -389,3 +389,64 @@ def test_task_stdout():
...
+        log.info("Job has re-registered")
+        sdk_tasks.check_running(job_service_name, 1)
+        log.info("Job has re-started")
+
+        restarted_service_info = shakedown.get_service(job_service_name).dict()
+        assert service_info['id'] != restarted_service_info['id'], "Job has restarted with same framework Id"
+    finally:
+        kill_info = utils.kill_driver(driver_id, utils.SPARK_SERVICE_NAME)
+        log.info("{}".format(kill_info))
+        kill_info = json.loads(kill_info)
```
Comment (akirillov) [inline tests/test_spark.py]: let's avoid reassignment of the variable to a value with a different type. we can do `loads` in the assert statement the line below

File: tests/test_spark.py [hunk 11]
```diff
@@ -389,3 +389,64 @@ def test_task_stdout():
...
+
+@pytest.mark.sanity
+def test_supervise_conflict_frameworkid():
+    job_service_name = "MockTaskRunner"
+
+    @retrying.retry(
+        wait_fixed=1000,
+        stop_max_delay=600 * 1000,
+        retry_on_result=lambda res: not res)
+    def wait_job_present(present):
```
Comment (akirillov) [inline tests/test_spark.py]: It looks like we keep reimplementing different flavours of this wait for each test, we should collect and extract all these waiting functions into a separate module. (not a part of this PR)

File: tests/test_spark.py [hunk 12]
```diff
@@ -389,3 +389,64 @@ def test_task_stdout():
...
+        service_info = shakedown.get_service(job_service_name).dict()
+        driver_regex = "spark.mesos.driver.frameworkId={}".format(service_info['id'])
+        status, stdout = shakedown.run_command_on_agent(service_info['hostname'], 
+                "ps axo pid,command | grep -v grep | grep '{}' | cut -d' ' -f1 | sudo xargs kill -9".format(driver_regex),
+                username=sdk_cmd.LINUX_USER)
+
+        if status:
+            log.info("Killed all the PIDs")
+        else:
+            log.info("Unable to kill PIDs")
```
Comment (akirillov) [inline tests/test_spark.py]: let's remove this conditional logging after switching to `sdk_cmd.kill_task_with_pattern`

File: tests/test_spark.py (hunk 9 above)
Comment (farhan5900) [reply tests/test_spark.py]: Simple `java` will not work, because one of the process is `docker run`. We can use `job_service_name` or only `frameworkId`.

File: tests/test_spark.py (hunk 8 above)
Comment (farhan5900) [reply tests/test_spark.py]: changed `shakedown.run_command_on_agent` to `sdk_cmd.kill_task_with_pattern`.

File: tests/test_spark.py (hunk 10 above)
Comment (farhan5900) [reply tests/test_spark.py]: moved it to the `assert` statement.

File: tests/test_spark.py (hunk 12 above)
Comment (farhan5900) [reply tests/test_spark.py]: Removed it.

File: tests/test_spark.py (hunk 1 above)
Comment (akirillov) [reply tests/test_spark.py]: @farhan5900, it seems that something is odd with the way the fixture works as pointed out by @samvantran. I'd suggest to do some debugging of the test itself and look at the cluster state at the moment of calling `utils.submit_job` by pausing the execution.

**Comment by akirillov** [comment]
@farhan5900, it turned out that the behaviour of foldered Spark tests led to environment modifications which prevented supervised tests (and a few others if executed multiple times on the same cluster) fail. TL;DR - those tests uninstalled CLI and not always cleaned up ZK. This PR addresses these issues: https://github.com/mesosphere/spark-build/pull/514, please rebase on master after it is merged.
